File: behaviorial/state/typescript/concrete-state.ts
import { AudioPlayer } from './context';
import { State } from './interface';

export class PlayingState implements State {
	play(player: AudioPlayer): void {
		console.log('Already playing.');
	}

	pause(player: AudioPlayer): void {
		console.log('Pausing the player.');
		player.setState(new PausedState());
	}

	stop(player: AudioPlayer): void {
		console.log('Stopping the player.');
		player.setState(new StoppedState());
	}
}

export class PausedState implements State {
	play(player: AudioPlayer): void {
		console.log('Resuming the player.');
		player.setState(new PlayingState());
	}

	pause(player: AudioPlayer): void {
		console.log('Already paused.');
	}

	stop(player: AudioPlayer): void {
		console.log('Stopping the player.');
		player.setState(new StoppedState());
	}
}

export class StoppedState implements State {
	play(player: AudioPlayer): void {
		console.log('Starting the player.');
		player.setState(new PlayingState());
	}

	pause(player: AudioPlayer): void {
		console.log('Cannot pause. Player is stopped.');
	}

	stop(player: AudioPlayer): void {
		console.log('Already stopped.');
	}
}

File: behaviorial/state/python/concrete_state.py


from interface import State


class PlayingState(State):
    def play(self, player):
        print("Already playing.")

    def pause(self, player):
        print("Pausing the player.")
        player.set_state(PausedState())

    def stop(self, player):
        print("Stopping the player.")
        player.set_state(StoppedState())


class PausedState(State):
    def play(self, player):
        print("Resuming the player.")
        player.set_state(PlayingState())

    def pause(self, player):
        print("Already paused.")

    def stop(self, player):
        print("Stopping the player.")
        player.set_state(StoppedState())


class StoppedState(State):
    def play(self, player):
        print("Starting the player.")
        player.set_state(PlayingState())

    def pause(self, player):
        print("Cannot pause. Player is stopped.")

    def stop(self, player):
        print("Already stopped.")
